book/listings/todo_app/1/window/mod.rs

```diff
@@ -0,0 +1,213 @@
+            // Get `TodoRow` from `ListItem`
+            let todo_row = list_item
+                .child()
+                .expect("The child has to exist.")
+                .downcast::<TodoRow>()
+                .expect("The child has to be a `TodoRow`.");
+
+            todo_row.unbind_item();
+        });
+
+        // Set the factory of the list view
+        let imp = imp::Window::from_instance(&self);
+        imp.list_view.set_factory(Some(&factory));
+    }
+
+    fn setup_callbacks(&self) {
+        // Get state
+        let imp = imp::Window::from_instance(&self);
+        let model = self.model();
+
+        // Setup callback so that activation of the entry
+        // creates a new todo object and clears the entry
+        imp.entry
+            .connect_activate(clone!(@weak model => move |entry| {
+                let buffer = entry.buffer();
+                let content = buffer.text();
+                let todo_object = TodoObject::new(content, false);
+                model
+                    .append(&todo_object);
+                buffer.set_text("");
+            }));
+
+        // Setup callback so that click on the clear_button
+        // removes all done tasks
+        imp.clear_button
+            .connect_clicked(clone!(@weak model => move |_| {
+                let mut position = 0;
+                while let Some(item) = model.item(position) {
+                    // Get `TodoObject` from `glib::Object`
+                    let todo_object = item
+                        .downcast_ref::<TodoObject>()
+                        .expect("The object needs to be of type `TodoObject`.");
+
+                    if todo_object.is_completed() {
+                        model.remove(position);
+                    } else {
+                        position += 1;
+                    }
+                }
+            }));
+    }
+
+    fn setup_shortcut_window(&self) {
+        // Get `ShortcutsWindow` via `gtk::Builder`
+        let builder = gtk::Builder::from_string(include_str!("shortcuts.ui"));
+        let shortcuts = builder
+            .object("shortcuts")
+            .expect("Could not get object `shortcuts` from builder.");
+
+        // After calling this method, calling the action "win.show-help-overlay" will show the shortcut window
+        self.set_help_overlay(Some(&shortcuts));
+    }
+
+    fn setup_filter_action(&self) {
+        // Get state
+        let imp = imp::Window::from_instance(&self);
+
+        // Create action from key "filter"
+        let filter_action = imp.settings.create_action("filter");
+
+        // Add action "filter" to action group "win"
+        self.add_action(&filter_action);
+    }
+
+    fn filter(&self) -> Option<CustomFilter> {
+        // Get state
+        let imp = imp::Window::from_instance(&self);
+
+        // Get value from settings
+        let value: String = imp.settings.get("filter");
+
+        let filter_all = CustomFilter::new(|_| true);
```

Comment:
Do you need to set a filter for all?

Reply:
What's the alternative?

Reply:
setting None as a filter?

Reply:
This is a function that returns a customfilter. That does not make much sense in that context.

Reply:
Ah now I get it.
You're completely right!